### PR TITLE
Linux snap fixes

### PR DESCRIPTION
**Pull request by AMZN-Phil**
Combined set of snap fixes for Linux including avoiding attempting to write files inside the read only snap. Adding additional search paths to when running from snap to search those directories first, and making sure the correct permissions are set.

### CODE REVIEW

**Review by amzn-changml — APPROVED** [review]
Pending a snap build from this branch

File: cmake/Platform/Linux/Packaging/snapcraft.yaml.in [hunk 1]
```diff
@@ -12,55 +12,72 @@ parts:
...
-     - libice6
-     - libsm6
-     - libcurl4
-     - libfontconfig1
      - libunwind-dev
-    stage-snaps:
-     - cmake
+     - pkg-config
+     - libc-dev
+     - libstdc++-12-dev
```
Comment (lumberyard-employee-dm) [inline cmake/Platform/Linux/Packaging/snapcraft.yaml.in:44]: So this is using clang14 with GNU libstdc++12?

File: scripts/o3de/o3de/manifest.py [hunk 2]
```diff
@@ -25,7 +25,10 @@
 # Directory methods
 
 def get_this_engine_path() -> pathlib.Path:
-    return pathlib.Path(os.path.realpath(__file__)).parents[3].resolve()
+    if "SNAP" in os.environ:
+        return pathlib.Path(os.environ.get('SNAP')) / os.environ.get('SNAP_BUILD')
+    else:
+        return pathlib.Path(os.path.realpath(__file__)).parents[3].resolve()
```
Comment (lumberyard-employee-dm) [inline scripts/o3de/o3de/manifest.py:31]: Why is SNAP specific logic being placed in non-snap code?

The manifest.py should be within the engine root even if it is in a SNAP.

File: cmake/Platform/Linux/Packaging/libc.so
Comment (AMZN-alexpete) [inline cmake/Platform/Linux/Packaging/libc.so:1]: did you mean to include this `libc.so` file?

File: cmake/Platform/Linux/Packaging/snapcraft.yaml.in [hunk 3]
```diff
@@ -12,55 +12,72 @@ parts:
...
+      find $SNAPCRAFT_STAGE/usr/lib/ -type f -executable -exec patchelf --set-interpreter /snap/core22/current/lib64/ld-linux-x86-64.so.2 {} \;
+      find $SNAPCRAFT_STAGE/usr/bin/ -type f -executable -exec patchelf --force-rpath --set-rpath $(patchelf --print-rpath {})':/snap/core22/current/lib/x86_64-linux-gnu:/snap/core22/current/usr/lib/x86_64-linux-gnu' {} \;
+      find $SNAPCRAFT_STAGE/usr/lib/ -type f -exec patchelf --force-rpath --set-rpath $(patchelf --print-rpath {})':/snap/core22/current/lib/x86_64-linux-gnu:/snap/core22/current/usr/lib/x86_64-linux-gnu' {} \;
+      find $SNAPCRAFT_STAGE/lib/ -type f -exec patchelf --force-rpath --set-rpath $(patchelf --print-rpath {} )':/snap/core22/current/lib/x86_64-linux-gnu:/snap/core22/current/usr/lib/x86_64-linux-gnu' {} \;
+      find $SNAPCRAFT_STAGE/lib32/ -type f -exec patchelf --force-rpath --set-rpath $(patchelf --print-rpath {} )':/snap/core22/current/lib/x86_64-linux-gnu:/snap/core22/current/usr/lib/x86_64-linux-gnu' {} \;
+      cp $SNAPCRAFT_STAGE/${CPACK_PACKAGE_VERSION}/cmake/Platform/Linux/Packaging/libc.so $SNAPCRAFT_STAGE/usr/lib/x86_64-linux-gnu/libc.so
```
Comment (AMZN-alexpete) [inline cmake/Platform/Linux/Packaging/snapcraft.yaml.in:58]: is there a better way to get libc.so than including it ourselves?

File: python/get_python.sh [hunk 4]
```diff
@@ -27,7 +27,17 @@ install_dependencies () {
...
     fi
 
-    $DIR/pip.sh install -e $DIR/../scripts/o3de --no-deps --disable-pip-version-check  --no-warn-script-location
+    # if we're building a container app, create a package from o3de then install that to remove absolute paths to o3de scripts
+    if [ $O3DE_BUILDING_CONTAINER == 1 ]; then
+        pushd $DIR/../scripts/o3de/
+        $DIR/python.sh setup.py sdist
+        popd
+
+        $DIR/pip.sh install $DIR/../scripts/o3de/dist/o3de-1.0.0.tar.gz --no-deps --disable-pip-version-check --no-cache
```
Comment (AMZN-alexpete) [inline python/get_python.sh:36]: should we be pulling the filename from an ENV - assuming the `1.0.0` is based on the engine version, should we get that from `engine.json` or are they unrelated?

File: python/get_python.sh [hunk 5]
```diff
@@ -27,7 +27,17 @@ install_dependencies () {
         return $retVal
     fi
 
-    $DIR/pip.sh install -e $DIR/../scripts/o3de --no-deps --disable-pip-version-check  --no-warn-script-location
+    # if we're building a container app, create a package from o3de then install that to remove absolute paths to o3de scripts
+    if [ $O3DE_BUILDING_CONTAINER == 1 ]; then
```
Comment (AMZN-alexpete) [inline python/get_python.sh:31]: could we just check if `$O3DE_PACKAGE_TYPE` is "SNAP"?

File: cmake/Platform/Linux/Configurations_linux.cmake [hunk 6]
```diff
@@ -63,4 +95,9 @@ endif()
 
 ly_set(CMAKE_BUILD_WITH_INSTALL_RPATH TRUE)
 ly_set(CMAKE_INSTALL_RPATH_USE_LINK_PATH FALSE)
-ly_set(CMAKE_INSTALL_RPATH "$ORIGIN")
+
+if ($ENV{O3DE_SNAP})
+    ly_set(CMAKE_INSTALL_RPATH "$ORIGIN:/snap/core22/current/lib/x86_64-linux-gnu:/snap/core22/current/usr/lib/x86_64-linux-gnu")
```
[the remark this answers is not in the document]
Comment (AMZN-Phil) [reply cmake/Platform/Linux/Configurations_linux.cmake:100]: It basically runs in a container, so will act as O3DE running on Ubuntu 22 on any distro that supports snap,  we shouldn't need to make any additional changes for Ubuntu 20 (the snap has been tested on 20.04 as well as 22.04, and also Amazon Linux with a custom snapd).

File: cmake/Platform/Linux/Packaging/snapcraft.yaml.in (hunk 3 above)
Comment (AMZN-Phil) [reply cmake/Platform/Linux/Packaging/snapcraft.yaml.in:58]: Not currently due to the standard one reaching out to the system causing some compilation problems when running the snap on a different version/distro than what the snap was built with.

File: python/get_python.sh (hunk 4 above)
Comment (AMZN-Phil) [reply python/get_python.sh:36]: They are unrelated, the is the version of the o3de scripts, we could version these the same as the engine version but 1.0.0 seems to be used across all platforms currently.

File: scripts/o3de/o3de/manifest.py (hunk 2 above)
Comment (AMZN-Phil) [reply scripts/o3de/o3de/manifest.py:31]: This was a workaround for a very specific issue I found with snapd causing this to return an incorrect path, there has been a couple of updates but haven't retested this with the latest version yet.

File: cmake/Platform/Linux/Packaging/snapcraft.yaml.in (hunk 3 above)
Comment (AMZN-alexpete) [reply cmake/Platform/Linux/Packaging/snapcraft.yaml.in:58]: I think we need to come up with a way not to include any pre-built binaries in this repository.   There are also legal implications to consider.

File: scripts/o3de/o3de/manifest.py (hunk 2 above)
Comment (AMZN-Phil) [reply scripts/o3de/o3de/manifest.py:31]: Yes, this was due to the path returned being incorrect when running from the snap container, it was returning a temporary folder and not a path to the o3de snap.

Comment (lumberyard-employee-dm) [reply scripts/o3de/o3de/manifest.py:31]: Thanks for the clarification.
Could you add that as a comment, that in a snap environment the `__file__` python variable is set to a temporary folder?

**Review by AMZN-alexpete — APPROVED** [review]

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by amzn-changml — APPROVED** [review]
Confirmed that the current changes to the PR branch creates a viable snap install. 

The `o3de` binary starts the editor, but is failing for a secondary issue (the packaged python runtime is not able to be used). That can be solved by another PR.

The bug for the secondary issue can be found here: https://github.com/o3de/o3de/issues/14902

**Pull request merged**